BUILD: Upload wheels to anaconda,org

mattip opened the pull request:
Fixes #20262

Upload wheels to anaconda.org via cibuildwheel. The logic is:
- pushing a new tag will trigger a guthub action with `github.ref='refs/tags/v*' which will upload the wheels to [multibuild-wheels-staging](https://anaconda.org/multibuild-wheels-staging/numpy)
- the cron-scheduled job or NumPy team members clicking on the "Run workflow" button from Actions (in the menu bar above) -> Wheel builder will upload wheels to [scipy-wheels-nightly](https://anaconda.org/scipy-wheels-nightly/)

This uses the secret tokens which were generated from https://anaconda.org/multibuild-wheels-staging/settings/access https://anaconda.org/scipy-wheels-nightly/settings/access and then added to this repo at https://github.com/numpy/numpy/settings/environments/231368784/edit

This will probably influence the [release instructions](https://numpy.org/devdocs/dev/releasing.html?highlight=howto_release#trigger-the-wheel-builds)

I tested this in my fork of the repo. The commit ref is echoed in the [Get commit message step](https://github.com/mattip/numpy/runs/4935793608?check_suite_focus=true) (open the "Get commit message" fold in the CI run area to see the temporary tag in my fork of `refs/tags/v2.0.1.dev0`. Then this causes "push and tag event" to be echoed from the appropriate branch of the `if` in the "Setup Upload Variables" fold in any of the wheel builds, which will set up the `$TOKEN` and `$ANACONDA_ORG` to be appropriately set.

The use of the secrets and final upload step should Just Work(tm) but may require some tweaks.

Thanks to @bashtage for the parallel work done in MacPython/statsmodels-wheels#97

@lithomas1 - does this look right?


Along the way, I also added `if: github.repository == 'numpy/numpy'` to some other workflows so they do not run on forks.

mattip committed:
BUILD: upload wheels to anaconda

mattip committed:
allow action to run

mattip committed:
run step on tag merge

mattip committed:
BUILD: do not run workflows on forks

mattip committed:
BUILD: echo github.ref

mattip committed:
do not run on forks

mattip committed:
BUILD: use secrets

mattip commented on .github/workflows/wheels.yml:
```diff
@@ -125,3 +127,44 @@ jobs:
+          elif [ "schedule" == "${{ github.event_name }}" ] || [ "workflow_dispatch" == "${{ github.event_name }}" ]; then
+            echo scheduled or dispatched event
+            echo "ANACONDA_ORG=scipy-wheels-nightly" >> $GITHUB_ENV
+            echo "TOKEN=${{ secrets.NUMPY_NIGHTLY_UPLOAD_TOKEN }}" >> $GITHUB_ENV
```
@bashtage I used `${{ secrets.NUMPY_TOKEN }}` here like elsewhere in this repo where you used `$NUMPY_TOKEN`. I wonder if it makes a difference?

lithomas1 commented on .github/workflows/wheels.yml:
```diff
@@ -125,3 +127,44 @@ jobs:
+            echo "ANACONDA_UPLOAD=false" >> $GITHUB_ENV
+          fi
+      - name: Upload wheels
+        if: ${{ always() }}
```
I don't think wheels should be uploaded if any of the build steps above fail.

mattip replied:
I will check for env.TOKEN.

mattip replied:
It works, the step was skipped.

lithomas1 commented:
Maybe it's better to wait for all wheels to be present before uploading wheels.
(The main ones that are missing are the aarch64 ones).

Uploading wheels from here would also mean stopping the numpy-wheels uploads, so it might not be ideal if some wheels are missing.

(P.S. can you append [wheel build] to your next commit so that the wheel builder triggers and we can see if the workflow works?)

mattip committed:
BUILD: tweak conditions for steps [wheel build]

mattip commented:
> Maybe it's better to wait for all wheels to be present before uploading wheels.
> (The main ones that are missing are the aarch64 ones).

Those wheels should come from the travis action, like in numpy-wheels which uses azure for x86 / macOS builds and travis for aarch64.

I think the proper order would be:
- get approval for this PR
- submit a PR to remove the cron job from the [numpy-wheels azure](https://github.com/MacPython/numpy-wheels/blob/eeafa52ccbd5dcf939578bc6d495edff850669c3/azure-pipelines.yml#L2) pipeline
- merge the two PRs simultaneously.

Then subsequent PRs can deal with aarch64 and travis.

mattip commented:
The wheel build run finished, it reports [non-dispatch event](https://github.com/numpy/numpy/runs/4949015880?check_suite_focus=true#step:6:19) when decided whether to upload or not.

mattip committed:
BUILD: [wheel build] increase verbosity when testing

mattip commented:
The intermittent PyPy38 windows failure will be fixed by the next release, which should happen in the next few weeks. See https://foss.heptapod.net/pypy/pypy/-/issues/3604

mattip commented:
We should filter out the dummy tag for starting a new branch (like `v1.22.0.dev0`) and not upload those wheels.

charris commented:
Does this skip uploading duplicate builds? That can be handy when a build fails for unrelated reasons and needs to be rerun. Github actions doesn't allow us to selectively rerun a build. Hmm, makes me wonder if there is a way to break things up so that that is possible.

mattip commented:
> Does this skip uploading duplicate builds

No, like the current numpy-wheel runs on Azure, the entire build can be rerun but individual wheel builds cannot, and any new upload will replace and older one with the same name.

bashtage commented:
If you upload a dupe to Anaconda it just overwrites  the previous daily that will have the same git hash and other descriptors.

mattip committed:
exclude the *dev0 dummy tag

mattip commented:
Tags that end with `dev0` will not upload.

charris commented:
> and any new upload will replace and older one with the same name.

Ah, you use `--force` instead of  `--skip`. Another thing that can happen is the tag needs to be changed to another commit, i.e., deleted and replaced when something needs to be fixed. How is that handled? I'd almost rather hit a button than rely on tags.

lithomas1 commented:
> > and any new upload will replace and older one with the same name.
> 
> Ah, you use `--force` instead of `--skip`. Another thing that can happen is the tag needs to be changed to another commit, i.e., deleted and replaced when something needs to be fixed. How is that handled? I'd almost rather hit a button than rely on tags.

Maybe we can make the uploading to the staging area based off another commit trigger.
(i.e. adding [wheel release] to a commit to upload to staging)

mattip commented:
> Another thing that can happen is the tag needs to be changed to another commit, i.e., deleted and replaced when something needs to be fixed. 

How common is this? I think changing tags is discouraged as an anti-practice. It would be better to liberally use release candidates, then re-tag the last release candidate with the release tag once everything is OK.

> Maybe we can make the uploading to the staging area based off another commit trigger.
(i.e. adding [wheel release] to a commit to upload to staging)

I would prefer to put this in as-is (and fix the inevitable problems with tokens and uploading) and if needed in a future PR we can add more features.

mattip committed:
BUILD: use --skip instead of --force when uploading, like numpy-wheels

mattip commented:
> Ah, you use --force instead of --skip

Changed to use `--skip` like the current numpy-wheels workflows.

charris commented:
> How common is this?

Typos happen :) That said, it may be that adding an empty commit for a new tag might be the best solution. I do tend to think of the release branch as my private preserve.

mattip commented:
> I do tend to think of the release branch as my private preserve.

Rightly so. I think `--skipped` should make changing the tags plausible, if not desirable.

charris commented:
> merge the two PRs simultaneously.

Might just have them on different days (Wed and Sat) and we can compare before disabling the numpy-wheels build.

mattip committed:
BUILD: schedule weekly builds for Wed

mattip commented:
@charris thoughts?

charris commented:
Let's just put this in, I don't see any reason to worry about multiple nightly uploads.

charris commented:
Thanks Matti.

charris commented:
There hasn't been a wheels upload to the nightly repo. I think that should have happened?

mattip commented:
Let's see, I pushed the button "Run workflow" on https://github.com/numpy/numpy/actions/workflows/wheels.yml.

mattip commented:
Ahh, indeed, something is wrong. The "Upload wheels" build step was not triggered from [this scheduled run](https://github.com/numpy/numpy/runs/5045932095?check_suite_focus=true), but it is printing "scheduled or dispatched event" in the "Set Upload Variables" step. The condition `if: env.TOKEN` is not being set :(

mattip commented:
fixed in PR #20996
